### PR TITLE
Media transformation class and different transformations based on subsets of channels

### LINKED ISSUE #712: Different media transformations across different channels
Following #632, it is very easy to apply different media transformations to different subsets of channels. 

Here is an example API:

Some functions in examples: 

```python
@dataclass
class MediaConfig:
    name: str
    columns: list[str]
    adstock: AdstockTransformation
    saturation: SaturationTransformation
    adstock_first: bool

def get_media_values(media_configs: list[MediaConfig]) -> list[str]: 
    """Get the different levels used in the configurations."""
    
def apply_media_transformations(
    media_data: pt.TensorLike, 
    media_configs: list[MediaConfig], 
    model: pm.Model,
) -> pt.TensorVariable:
    """Apply the media transformations to subsets according to media_configs."""

```

Then using it could look like this:

```python
media_configs: list[MediaConfig]= [
    MediaConfig(
        name="offline",
        columns=["TV", "Radio"],
        adstock=GeometricAdstock(l_max=15),
        saturation=HillSaturation(),
        adstock_first=True,
    ),
    MediaConfig(
        name="online",
        columns=["Facebook", "Instagram", "YouTube", "TikTok"],
        adstock=GeometricAdstock(l_max=10),
        saturation=MichaelisMentenSaturation(),
        adstock_first=False,
    ),
]

media_values = get_media_values(media_configs)

coords = {
    "time": dates,
    "media": media_values,
}
model = pm.Model(coords=coords)

df: pd.DataFrame = ...  # Some data with all the columns

with model:
    media_data = pm.Data(
        "media_data",
        df.loc[:, media_values].to_numpy(),
        dims=("time", "media"),
    )
    media_transformation_data = pm.Deterministic(
        "channel_contributions",
        apply_media_transformation(media_data, media_configs, model),
        dims=("time", "media"),
    )
```

I would be interested in hearing if others are interested in this functionality.

### PR DESCRIPTION
**Pull request by wd60622**
## Description

Allow for media transformation on subset of channels

## Related Issue

- [x] Closes #712 
- [ ] Related to #

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [x] MMM
- [ ] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--968.org.readthedocs.build/en/968/

### CODE REVIEW

File: pymc_marketing/mmm/__init__.py [hunk 1]
```diff
@@ -39,13 +39,25 @@
 )
 from pymc_marketing.mmm.delayed_saturated_mmm import MMM, DelayedSaturatedMMM
 from pymc_marketing.mmm.fourier import MonthlyFourier, YearlyFourier
+from pymc_marketing.mmm.media_transformation import (
+    MediaConfig,
+    MediaConfigs,
```
Comment (juanitorduz) [inline pymc_marketing/mmm/__init__.py:44]: Shall we have more different names? Just as "s" at the end is a typo-rick hehe. For example `MediaConfig` and `MediaConfigList`?

File: pymc_marketing/mmm/media_transformation.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        ),
+        MediaConfig(
+            name="online",
+            columns=["Facebook", "Instagram", "YouTube", "TikTok"],
+            media_transformation=online_media_transform,
+        ),
+    ]
+
+
+Apply the media transformation to media data in PyMC model:
```
Comment (juanitorduz) [inline pymc_marketing/mmm/media_transformation.py:64]: Can this configuration specification be used through the `media_config` parameter with the `MMM` model class? If so, could you provide an example here as well 🙏 ?

**Review by juanitorduz — COMMENTED** [review]
Wow! Super Cool! Just left some questions.

File: pymc_marketing/mmm/__init__.py (hunk 1 above)
Comment (wd60622) [reply pymc_marketing/mmm/__init__.py:44]: Good idea 😆

File: pymc_marketing/mmm/media_transformation.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+    name: str
+    columns: list[str]
+    media_transformation: MediaTransformation
+
+
+MediaConfigs = list[MediaConfig]
+
+
+def apply_media_transformation(
```
Comment (wd60622) [inline pymc_marketing/mmm/media_transformation.py:211]: If these are wrapped in a class, then it should be pretty easy to make a consistent API with the `apply` method. 

WDYT?

File: pymc_marketing/mmm/media_transformation.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+    def __post_init__(self):
+        """Set the first and second transformations based on the adstock_first flag."""
+        self.first, self.second = (
+            (self.adstock, self.saturation)
+            if self.adstock_first
+            else (self.saturation, self.adstock)
+        )
+
+    def __call__(self, x, dim):
```
Comment (wd60622) [inline pymc_marketing/mmm/media_transformation.py:139]: I use __call__ here instead of an `apply` method used in individual transformations. WYDT?
As mentioned below, we can have 

```python
media_data = ...
online_media_transformation = MediaTransformation(adstock=..., saturation=..., ...)
offline_media_transformation = MediaTransformation(adstock=..., saturation=..., ...)
media_configs = [MediaConfig(media_transformation=online_media_transformation), MediaConfig(media_transformation=offline_media_transformation)]
comb_media_transformation = CombinedMediaTransformation(media_configs=media_configs)

# Make all three work the same
offline_media_transformation(media_data, dims="media")
online_media_transformation(media_data, dims="media")
comb_media_transformation(media_data, dims="media")
```

This would have have many conveniences

File: pymc_marketing/mmm/media_transformation.py (hunk 2 above)
Comment (wd60622) [reply pymc_marketing/mmm/media_transformation.py:64]: There isn't a `media_config` parameter at the moment. But we could make one 👀

File: pymc_marketing/mmm/media_transformation.py (hunk 3 above)
Comment (juanitorduz) [reply pymc_marketing/mmm/media_transformation.py:211]: yes!

File: pymc_marketing/mmm/media_transformation.py (hunk 4 above)
Comment (juanitorduz) [reply pymc_marketing/mmm/media_transformation.py:139]: Nice! 😎

**Comment by wd60622** [comment]
I've addressed all the feedback @juanitorduz  except for the integration with `MMM`. That would require a handful of adjustments and discussion. 

For instance, 
is this a new parameter for initialization? 

```python
media_transformation = MediaTransformation(...)
mmm = MMM(media_transformation=media_transformation, ...)

media_transformation = MediaConfigList(...)
mmm = MMM(media_transformation=media_transformation, ...)
```

That has some implications: 

- what do we do with current parameters `adstock`, `saturation`, `adstock_first`? Deprecate?
- How do we handle all of the model_config parameters? Should we store all of ModelTransformation / ModelConfigList there even though that is duplicated since there are stored in the individual transformations themselves
- Various plotting functions need to be rewritten in order to support multiple adstocks / saturations etc
- Budget optimizer will only conditionally work if single MediaTransformation

Definitely doable though

Do you want to do now or wait on this? Let me know how you imagine this being added in

**Review by juanitorduz — APPROVED** [review]
Yay! We should work on iterstions so let's merge this one and leave the MMM integration for another PR (can you please create an issue so that we keep track?)

**Comment by wd60622** [comment]
> Yay! We should work on iterstions so let's merge this one and leave the MMM integration for another PR (can you please create an issue so that we keep track?)

Yes, will make some issues now. 

Would you like me to increase the test coverage?

**Comment by juanitorduz** [comment]
@wd60622 from my side it looks ok! If you wanna test edge cases go for it.

**Comment by wd60622** [comment]
> @wd60622 from my side it looks ok! If you wanna test edge cases go for it. 

I will create an issue for it
